pkg/bot/teams_cloud.go

```diff
@@ -211,34 +218,32 @@ func (b *CloudTeams) start(ctx context.Context) error {
 func (b *CloudTeams) withRetries(ctx context.Context, log logrus.FieldLogger, maxRetries int, fn func() error) error {
 	b.failuresNo = 0
 	var lastFailureTimestamp time.Time
+	resettableBackoff := func(n uint, err error, cfg *retry.Config) time.Duration {
+		if !lastFailureTimestamp.IsZero() && time.Since(lastFailureTimestamp) >= successIntervalDuration {
+			// if the last run was long enough, we treat is as success, so we reset failures
+			log.Infof("Resetting failures counter as last failure was more than %s ago", successIntervalDuration)
+			b.failuresNo = 0
+		}
+		lastFailureTimestamp = time.Now()
+		b.failuresNo++
+		b.setFailureReason(health.FailureReasonConnectionError)
+
+		return retry.BackOffDelay(uint(b.failuresNo), err, cfg)
+	}
 	return retry.Do(
 		func() error {
 			err := fn()
-			if err != nil {
-				if !lastFailureTimestamp.IsZero() && time.Since(lastFailureTimestamp) >= successIntervalDuration {
-					// if the last run was long enough, we treat is as success, so we reset failures
-					log.Infof("Resetting failures counter as last failure was more than %s ago", successIntervalDuration)
-					b.failuresNo = 0
-				}
-
-				if b.failuresNo >= maxRetries {
-					b.setFailureReason(health.FailureReasonMaxRetriesExceeded)
-					log.Debugf("Reached max number of %d retries: %s", maxRetries, err)
-					return retry.Unrecoverable(err)
-				}
-
-				lastFailureTimestamp = time.Now()
-				b.failuresNo++
-				b.setFailureReason(health.FailureReasonConnectionError)
-				return err
+			if b.failuresNo >= maxRetries {
+				b.setFailureReason(health.FailureReasonMaxRetriesExceeded)
+				log.Debugf("Reached max number of %d retries: %s", maxRetries, err)
+				return retry.Unrecoverable(err)
 			}
-			b.setFailureReason("")
-			return nil
+			return err
 		},
 		retry.OnRetry(func(_ uint, err error) {
 			log.Warnf("Retrying Cloud Teams startup (attempt no %d/%d): %s", b.failuresNo, maxRetries, err)
 		}),
-		retry.Delay(retryDelay),
+		retry.DelayType(resettableBackoff),
 		retry.Attempts(0), // infinite, we cancel that by our own
 		retry.LastErrorOnly(true),
 		retry.Context(ctx),
```

Comment:
should we also add something similar here to

```go
	b.failuresNo = 0 // Reset the failures to start exponential back-off from the beginning
	b.setFailureReason("")
	b.log.Info("Botkube connected to Slack!")
```

?

Reply:
I couldn't get the point here, could you please elaborate a bit more?